File: android/src/main/java/au/com/touchline/basicnfc/BasicNFCPlugin.java
package au.com.touchline.basicnfc;

import java.util.concurrent.atomic.AtomicBoolean;
import android.os.Handler;
import android.os.Looper;

import android.nfc.NdefMessage;
import android.nfc.NdefRecord;
import android.nfc.Tag;
import android.nfc.NfcAdapter;
import android.nfc.tech.NdefFormatable;
import android.nfc.tech.Ndef;

import com.getcapacitor.JSObject;
import com.getcapacitor.Plugin;
import com.getcapacitor.PluginCall;
import com.getcapacitor.PluginMethod;
import com.getcapacitor.annotation.CapacitorPlugin;

import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;

import java.nio.charset.Charset;
import android.util.Log;

@CapacitorPlugin(name = "BasicNFC")
public class BasicNFCPlugin extends Plugin {
	private BasicNFC implementation = new BasicNFC();
	
	@PluginMethod
	public void echo(PluginCall call) {
		String value = call.getString("value");
		
		JSObject ret = new JSObject();
		ret.put("value", implementation.echo(value));
		call.resolve(ret);
	}
	
	// Method to process NFC messages from intents
	public void processNfcMessage(NdefMessage ndefMessage, String tagID) {
		if (ndefMessage != null) {
			NdefRecord[] records = ndefMessage.getRecords();
			JSONArray recordsArray = new JSONArray();
			
			for (NdefRecord record : records) {
				try {
					JSONObject recordJson = new JSONObject();
					recordJson.put("tnf", record.getTnf());
					recordJson.put("type", android.util.Base64.encodeToString(record.getType(), android.util.Base64.NO_WRAP));
					recordJson.put("id", android.util.Base64.encodeToString(record.getId(), android.util.Base64.NO_WRAP));
					recordJson.put("payload", android.util.Base64.encodeToString(record.getPayload(), android.util.Base64.NO_WRAP));
					recordsArray.put(recordJson);
				} catch (JSONException e) {
					Log.e("NFC", "Error converting record to JSON", e);
				}
			}
			
			JSObject ret = new JSObject();
			ret.put("records", recordsArray.toString());
			ret.put("tagID", tagID);
			
			// Notify listeners with the JSON data
			notifyListeners("nfcRawDataReceived", ret);
		}
	}
	
	private static final int TIMEOUT = 5000; // 5 seconds timeout
	
	@PluginMethod
	public void writeNFC(PluginCall call) {
		String message = call.getString("message"); // Assuming you're passing a string to write
		
		if (message == null) {
			call.reject("Message cannot be null");
			return;
		}
		
		if (NfcAdapter.getDefaultAdapter(getContext()) != null) {
			final NfcAdapter nfcAdapter = NfcAdapter.getDefaultAdapter(getContext());
			final AtomicBoolean writeSuccess = new AtomicBoolean(false);
			
			NfcAdapter.ReaderCallback readerCallback = new NfcAdapter.ReaderCallback() {
				@Override
				public void onTagDiscovered(Tag tag) {
					writeNdefMessage(tag, message, call);
					writeSuccess.set(true); // Signal that the write was successful
				}
			};
			
			// Enable reader mode for writing NFC
			nfcAdapter.enableReaderMode(getActivity(), readerCallback, NfcAdapter.FLAG_READER_NFC_A | NfcAdapter.FLAG_READER_NFC_B, null);
			
			// Schedule to disable reader mode after timeout
			final Handler handler = new Handler(Looper.getMainLooper());
			handler.postDelayed(new Runnable() {
				@Override
				public void run() {
					nfcAdapter.disableReaderMode(getActivity());
					
					if (!writeSuccess.get()) { // If the write hasn't happened yet
						call.reject("Timeout: No NFC tag detected");
					}
				}
			}, TIMEOUT);
		} else {
			call.reject("NFC is not supported on this device.");
		}
	}
	
	private void writeNdefMessage(Tag tag, String message, PluginCall call) {
		Ndef ndef = Ndef.get(tag);
		
		try {
			if (ndef == null) {
				// Let's try to format the tag if it's not NDEF compatible
				NdefFormatable format = NdefFormatable.get(tag);
				
				if (format != null) {
					format.connect();
					format.format(createNdefMessage(message));
					format.close();
					
					JSObject ret = new JSObject();
					ret.put("result", "NFC tag formatted and data written");
					call.resolve(ret);
					return;
				} else {
					call.reject("Tag is not NDEF capable and cannot be formatted.");
					return;
				}
			}
			
			ndef.connect();
			
			if (ndef.isWritable()) {
				ndef.writeNdefMessage(createNdefMessage(message));
				JSObject ret = new JSObject();
				ret.put("result", "Data written to NFC tag");
				call.resolve(ret);
			} else {
				call.reject("Tag is read-only");
			}
			
			ndef.close();
		} catch (Exception e) {
			call.reject("Error writing to NFC tag: " + e.getMessage());
		}
	}
	
	private NdefMessage createNdefMessage(String message) {
		byte[] langBytes = "en".getBytes(Charset.forName("US-ASCII"));
		byte[] textBytes = message.getBytes(Charset.forName("UTF-8"));
		byte[] payload = new byte[1 + langBytes.length + textBytes.length];
		payload[0] = (byte) langBytes.length;
		
		System.arraycopy(langBytes, 0, payload, 1, langBytes.length);
		System.arraycopy(textBytes, 0, payload, 1 + langBytes.length, textBytes.length);
		
		NdefRecord rtdTextRecord = new NdefRecord(
			NdefRecord.TNF_WELL_KNOWN, 
			NdefRecord.RTD_TEXT, 
			new byte[0], 
			payload
		);
		
		return new NdefMessage(new NdefRecord[]{rtdTextRecord});
	}
	
	@PluginMethod
	public void scanNFC(PluginCall call) {
		// Get the simulatePayload parameter from the call, defaulting to empty string if not provided
		String simulatePayload = call.getString("simulatePayload", "");
		
		if (NfcAdapter.getDefaultAdapter(getContext()) != null) {
			final NfcAdapter nfcAdapter = NfcAdapter.getDefaultAdapter(getContext());
			final AtomicBoolean scanSuccess = new AtomicBoolean(false);
			
			NfcAdapter.ReaderCallback readerCallback = new NfcAdapter.ReaderCallback() {
				@Override
				public void onTagDiscovered(Tag tag) {
					Ndef ndef = Ndef.get(tag);
					
					if (ndef != null) {
						try {
							ndef.connect();
							NdefMessage ndefMessage = ndef.getNdefMessage();
							ndef.close();
							
							if (ndefMessage != null) {
								// Convert the NDEF message to JSON and send back
								JSObject ret = new JSObject();
								ret.put("messages", convertNdefMessageToJson(ndefMessage));
								ret.put("tagID", android.util.Base64.encodeToString(tag.getId(), android.util.Base64.NO_WRAP));
								call.resolve(ret);
							} else {
								call.reject("Error: Tag has no data");
							}
							
							scanSuccess.set(true); // Signal that scanning was successful
						} catch (Exception e) {
							call.reject("Error: " + e.getMessage());
						}
					} else {
						call.reject("Error: Tag does not support NDEF");
					}
					
					// Disable reader mode after processing the tag
					nfcAdapter.disableReaderMode(getActivity());
				}
			};
			
			// Enable reader mode for scanning NFC
			nfcAdapter.enableReaderMode(getActivity(), readerCallback, NfcAdapter.FLAG_READER_NFC_A | NfcAdapter.FLAG_READER_NFC_B, null);
			
			// Set a timeout if you want to end scanning after a certain period
			final Handler handler = new Handler(Looper.getMainLooper());
			
			handler.postDelayed(new Runnable() {
				@Override
				public void run() {
					nfcAdapter.disableReaderMode(getActivity());
					
					if (!scanSuccess.get()) { // If no tag was detected
						if (!simulatePayload.isEmpty()) {
							try {
								// Create a fake NDEF record with the simulatePayload
								NdefRecord textRecord = NdefRecord.createTextRecord("en", simulatePayload);
								NdefMessage fakeMessage = new NdefMessage(new NdefRecord[]{textRecord});
								
								// Simulate a successful scan
								JSObject ret = new JSObject();
								ret.put("messages", convertNdefMessageToJson(fakeMessage));
								// Simulate a fake tag ID (you can customize this as needed)
								String fakeTagId = android.util.Base64.encodeToString("SIMULATED_TAG".getBytes(), android.util.Base64.NO_WRAP);
								ret.put("tagID", fakeTagId);
								call.resolve(ret);
							} catch (Exception e) {
								call.reject("Error simulating payload: " + e.getMessage());
							}
						} else {
							call.reject("timeout");
						}
					}
				}
			}, TIMEOUT); // Using the existing TIMEOUT constant
		} else {
			if (!simulatePayload.isEmpty()) {
				try {
					// Create a fake NDEF record with the simulatePayload
					NdefRecord textRecord = NdefRecord.createTextRecord("en", simulatePayload);
					NdefMessage fakeMessage = new NdefMessage(new NdefRecord[]{textRecord});
					
					// Simulate a successful scan
					JSObject ret = new JSObject();
					ret.put("messages", convertNdefMessageToJson(fakeMessage));
					// Simulate a fake tag ID (you can customize this as needed)
					String fakeTagId = android.util.Base64.encodeToString("SIMULATED_TAG".getBytes(), android.util.Base64.NO_WRAP);
					ret.put("tagID", fakeTagId);
					call.resolve(ret);
				} catch (Exception e) {
					call.reject("Error simulating payload: " + e.getMessage());
				}
			} else {
				call.reject("Error: NFC is not supported on this device.: " + simulatePayload);
			}
		}
	}
	
	private String convertNdefMessageToJson(NdefMessage ndefMessage) {
		JSONArray recordsArray = new JSONArray();
		
		for (NdefRecord record : ndefMessage.getRecords()) {
			try {
				JSONObject recordJson = new JSONObject();
				recordJson.put("tnf", record.getTnf());
				recordJson.put("type", android.util.Base64.encodeToString(record.getType(), android.util.Base64.NO_WRAP));
				recordJson.put("id", android.util.Base64.encodeToString(record.getId(), android.util.Base64.NO_WRAP));
				recordJson.put("payload", android.util.Base64.encodeToString(record.getPayload(), android.util.Base64.NO_WRAP));
				recordsArray.put(recordJson);
			} catch (JSONException e) {
				Log.e("NFC", "Error converting record to JSON", e);
			}
		}
		
		return recordsArray.toString();
	}
}

File: ios/Sources/BasicNFCPlugin/BasicNFCPlugin.swift
import Capacitor
import CoreNFC

@objc(BasicNFCPlugin)
public class BasicNFCPlugin: CAPPlugin, NFCNDEFReaderSessionDelegate {
	private var nfcSession: NFCNDEFReaderSession?
	private var currentCall: CAPPluginCall?
	private let timeout: TimeInterval = 5.0 // 5 seconds timeout
	
	@objc func echo(_ call: CAPPluginCall) {
		guard let value = call.getString("value") else {
			call.reject("Value cannot be null")
			return
		}
		
		let result = ["value": value]
		call.resolve(result)
	}
	
	@objc func writeNFC(_ call: CAPPluginCall) {
		guard let message = call.getString("message") else {
			call.reject("Message cannot be null")
			return
		}
		
		guard NFCNDEFReaderSession.readingAvailable else {
			call.reject("NFC is not supported on this device")
			return
		}
		
		currentCall = call
		nfcSession = NFCNDEFReaderSession(delegate: self, queue: nil, invalidateAfterFirstRead: true)
		nfcSession?.alertMessage = "Hold your iPhone near an NFC tag to write data."
		nfcSession?.begin()
		
		// Timeout handling
		DispatchQueue.main.asyncAfter(deadline: .now() + timeout) { [weak self] in
			if let session = self?.nfcSession, session.isReady {
				session.invalidate(errorMessage: "Timeout: No NFC tag detected")
				self?.currentCall?.reject("Timeout: No NFC tag detected")
				self?.currentCall = nil
			}
		}
	}
	
	@objc func scanNFC(_ call: CAPPluginCall) {
		let simulatePayload = call.getString("simulatePayload") ?? ""
		
		guard NFCNDEFReaderSession.readingAvailable else {
			if !simulatePayload.isEmpty {
				// Simulate NFC scan
				let result = self.simulateNFCScan(payload: simulatePayload)
				call.resolve(result)
			} else {
				call.reject("NFC is not supported on this device")
			}
			return
		}
		
		currentCall = call
		nfcSession = NFCNDEFReaderSession(delegate: self, queue: nil, invalidateAfterFirstRead: true)
		nfcSession?.alertMessage = "Hold your iPhone near an NFC tag to scan."
		nfcSession?.begin()
		
		// Timeout handling
		DispatchQueue.main.asyncAfter(deadline: .now() + timeout) { [weak self] in
			if let session = self?.nfcSession, session.isReady {
				if !simulatePayload.isEmpty {
					let result = self?.simulateNFCScan(payload: simulatePayload) ?? [:]
					call.resolve(result)
				} else {
					session.invalidate(errorMessage: "Timeout: No NFC tag detected")
					self?.currentCall?.reject("timeout")
				}
				self?.currentCall = nil
			}
		}
	}
	
	// MARK: - NFCNDEFReaderSessionDelegate
	
	public func readerSession(_ session: NFCNDEFReaderSession, didInvalidateWithError error: Error) {
		if let call = currentCall {
			call.reject("NFC session error: \(error.localizedDescription)")
			currentCall = nil
		}
	}
	
	public func readerSession(_ session: NFCNDEFReaderSession, didDetectNDEFMessages messages: [NFCNDEFMessage]) {
		guard let call = currentCall else { return }
		
		if call.callbackId == "writeNFC" {
			// Writing operation
			guard let message = call.getString("message"),
				let ndefMessage = createNDEFMessage(from: message) else {
				call.reject("Invalid message format")
				return
			}
			
			session.connect(to: messages[0].records[0].tag!) { error in
				if let error = error {
					call.reject("Failed to connect to tag: \(error.localizedDescription)")
					return
				}
				
				let tag = messages[0].records[0].tag as! NFCNDEFTag
				tag.writeNDEF(ndefMessage) { error in
					if let error = error {
						call.reject("Failed to write to tag: \(error.localizedDescription)")
					} else {
						call.resolve(["result": "Data written to NFC tag"])
					}
					self.currentCall = nil
				}
			}
		} else {
			// Scanning operation
			let result = convertNDEFMessagesToJSON(messages: messages)
			call.resolve(result)
			currentCall = nil
		}
	}
	
	// MARK: - Helper Methods
	
	private func createNDEFMessage(from text: String) -> NFCNDEFMessage? {
		guard let textRecord = NFCNDEFRecord(payload: text.data(using: .utf8) ?? Data(),
										typeNameFormat: .nfcWellKnown,
										type: "T".data(using: .utf8) ?? Data()) else {
			return nil
		}
		return NFCNDEFMessage(records: [textRecord])
	}
	
	private func convertNDEFMessagesToJSON(messages: [NFCNDEFMessage]) -> [String: Any] {
		var recordsArray: [[String: Any]] = []
		
		for message in messages {
			for record in message.records {
				let recordDict: [String: Any] = [
					"tnf": record.typeNameFormat.rawValue,
					"type": record.type.base64EncodedString(),
					"id": record.identifier.base64EncodedString(),
					"payload": record.payload.base64EncodedString()
				]
				recordsArray.append(recordDict)
			}
		}
		
		return [
			"messages": recordsArray,
			"tagID": messages.first?.records.first?.identifier.base64EncodedString() ?? ""
		]
	}
	
	private func simulateNFCScan(payload: String) -> [String: Any] {
		guard let textRecord = NFCNDEFRecord(payload: payload.data(using: .utf8) ?? Data(),
										typeNameFormat: .nfcWellKnown,
										type: "T".data(using: .utf8) ?? Data()),
			let fakeMessage = NFCNDEFMessage(records: [textRecord]) else {
			return [:]
		}
		
		return convertNDEFMessagesToJSON(messages: [fakeMessage])
	}
}
